### PR TITLE
Add ability to configure specific recommendations during BotKube installation/upgrade

### PR DESCRIPTION
**Pull request by pkosiec**
## Description

Changes proposed in this pull request:

- Add ability to configure recommendations in granular way inside the config
- Nest `recommendations` under Kubernetes source (`sources.'alias'.kubernetes`)
- Post recommendations according to source configurations (full support will be introduced in https://github.com/kubeshop/botkube/issues/676)
- Move recommendations out of filters
- Add unit tests for all recommendations
- Improve recommendation messages
- Update E2E tests
- Fix receiving events on other namespaces
- Remove unused code

## Limitations

Currently, recommendations are not send if someone disables watching for Pod `create` and Ingress `create` events, even if all recommendations toggles are set to `true`. This is a behavior known, but will be fixed in a follow-up as a part of https://github.com/kubeshop/botkube/issues/676 or after (if it's complexity will be bigger).

## Testing

Install BotKube from this PR with all/part of the recommendations enabled:

Overrides:
```
communications:
  'default-group':
    slack:
      enabled: true
      token: {TOKEN}
      channels:
        'default':
          name: botkube-test
          bindings:
            executors:
              - kubectl-read-only
            sources:
              - k8s-events
settings:
  clusterName: dev
image:
  pullPolicy: Always
  repository: kubeshop/pr/botkube
  tag: 687-PR
```
Command:
```
helm install botkube --namespace botkube ./helm/botkube -f /tmp/values.yaml --wait
```

Apply the following YAML:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: failing
spec:
  containers:
  - name: nginx
    image: nginx:latest
    command: ["foo"]
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: ingress
spec:
  tls:
    - hosts:
        - foo.example.com
      secretName: foo
  rules:
    - host: "foo"
      http:
        paths:
          - path: "/testpath"
            pathType: Exact
            backend:
              service:
                name: "foo"
                port:
                  number: 3001
```

Now, turn some recommendations off during installation and apply the YAML again. 

## Related issue(s)

#684

### CODE REVIEW

File: pkg/recommendation/factory.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	}
+	return newRecommendationsSet(f.logger, set)
+}
+
+func (f *Factory) createIfNotExist(condition bool, set map[string]Recommendation, constructorFn func() Recommendation) {
+	if !condition {
+		return
+	}
+
+	// this solution has a downside that we need to create an instance every time,
```
Comment (mszostok) [inline pkg/recommendation/factory.go:61]: **nit**, you can just use your own `tag` for `createIfNotExist` func:

```go
const ingressTLSRec = "ingres-tls-rec"
		f.createIfNotExist(ingressTLSRec, source.Recommendations.Ingress.TLSSecretValid, set,
			func() Recommendation { return NewIngressTLSSecretValid(f.dynamicCli) },
		)
```
or even
```go
const (
	IngressTLS Rec = iota
	PodLabels
	// ....    
)
```

File: pkg/recommendation/factory_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/kubeshop/botkube/pkg/config"
+	"github.com/kubeshop/botkube/pkg/recommendation"
+)
+
+func TestFactory_NewForSources(t *testing.T) {
+	// given
+	sources := config.IndexableMap[config.Sources]{
```
Comment (mszostok) [inline pkg/recommendation/factory_test.go:16]: we should stop using the `config.IndexableMap` in places where `GetFirst()` is not needed.

File: pkg/recommendation/ingress_tls_secret_valid.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package recommendation
```
Comment (mszostok) [inline pkg/recommendation/ingress_tls_secret_valid.go:1]: I assume that the code logic was not changed, right? there is no need to "recheck" it?

**Review by mszostok — APPROVED** [review]
This PR LGTM as we will address issues in the follow-up. Current issues/limitations:
- if `recommendations` are enabled but the `kubernetes.resources` are not specified, then it won't work
- the `recommendations` are a separate "source" but it will be moved under `kubernetes` to make them more BotKube specific and associate directly with K8s as we are not sure about the final approach to "generic" recommendations.

File: pkg/recommendation/ingress_tls_secret_valid.go (hunk 3 above)
Comment (pkosiec) [reply pkg/recommendation/ingress_tls_secret_valid.go:1]: Nope, I just simplified / slightly improved the code overall without changing the way how it works.

For Pod Labels Set I got rid of one snippet which wasn't necessary.

I tested all recommendations manually + implemented unit tests as well.

File: pkg/recommendation/factory.go (hunk 1 above)
Comment (pkosiec) [reply pkg/recommendation/factory.go:61]: If you don't mind I would keep it as it is as for me is less error-prone aporoach from the extensibility point of view.

**Pull request merged**